token/program/src/pack.rs

```diff
@@ -0,0 +1,79 @@
+//! State transition types
+
+use crate::error::TokenError;
+use solana_sdk::program_error::ProgramError;
+
+/// Check is a token state is initialized
+pub trait IsInitialized {
+    /// Is initialized
+    fn is_initialized(&self) -> bool;
+}
+
+/// Depends on Sized
+pub trait Sealed: Sized {}
+
+/// Safely and efficiently (de)serialize account state
+pub trait Pack: Sealed {
+    /// The length, in bytes, of the packed representation
+    const LEN: usize;
+    #[doc(hidden)]
+    fn pack_into_slice(&self, dst: &mut [u8]);
+    #[doc(hidden)]
+    fn unpack_from_slice(src: &[u8]) -> Result<Self, ProgramError>;
+
+    /// Borrow `Self` from `input` for the duration of the call to `f`, but first check that `Self`
+    /// is initialized
+    #[inline(never)]
+    fn unpack_mut<U>(
+        input: &mut [u8],
+        f: &mut dyn FnMut(&mut Self) -> Result<U, ProgramError>,
```

Comment:
Why the trait object here and not `&mut F where F: FnMut(&mut Self) -> Result<U, ProgramError>`?

Reply:
That'll work too, updated